docs/reference/deploy/master-config-reference.rst
.. _master-config-reference:

################################
 Master Configuration Reference
################################

The Determined master supports various configuration settings that can be set via a :ref:`YAML
configuration file <topic-guides_yaml>`, environment variables, or command-line options. The
configuration file is typically located at ``/etc/determined/master.yaml`` on the master and it is
read when the master starts up.

To inspect the configuration of an active master, use the Determined CLI and execute the command
``det master config``.

The master supports the following configuration settings:

*****************
 ``config_file``
*****************

Path to the master configuration file. Normally this should only be set via an environment variable
or command-line option. Defaults to ``/etc/determined/master.yaml``.

**********
 ``port``
**********

The TCP port on which the master accepts incoming connections. If TLS has been enabled, defaults to
``8443``; otherwise defaults to ``8080``.

.. _master-task-container-defaults:

*****************************
 ``task_container_defaults``
*****************************

Specifies defaults for all task containers. A task represents a single schedulable unit, such as a
trial, command, or TensorBoard.

``shm_size_bytes``
==================

The size (in bytes) of ``/dev/shm`` for Determined task containers. Defaults to ``4294967296``.

``network_mode``
================

The Docker network to use for the Determined task containers. If this is set to ``host``, `Docker
host-mode networking <https://docs.docker.com/network/drivers/host/>`__ will be used instead.
Defaults to ``bridge``.

.. _master-config-reference-dtrain-network-interface:

``dtrain_network_interface``
============================

The network interface to use during distributed training. If not set, Determined automatically
determines the network interface to use.

When training a model with multiple machines, the host network interface used by each machine must
have the same interface name across machines. The network interface to use can be determined
automatically, but there may be issues if there is an interface name common to all machines but it
is not routable between machines. Determined already filters out common interfaces like ``lo`` and
``docker0``, but agent machines may have others. If interface detection is not finding the
appropriate interface, the ``dtrain_network_interface`` option can be used to set it explicitly
(e.g., ``eth11``).

.. include:: ../../_shared/note-dtrain-learn-more.txt

``cpu_pod_spec``
================

Defines the default pod spec which will be applied to all CPU-only tasks when running on Kubernetes.
See :ref:`custom-pod-specs` for details.

``gpu_pod_spec``
================

Defines the default pod spec which will be applied to all GPU tasks when running on Kubernetes. See
:ref:`custom-pod-specs` for details.

``image``
=========

Defines the default Docker image to use when executing the workload. If a Docker image is specified
in the :ref:`experiment config <exp-environment-image>` this default is overridden. This image must
be accessible via ``docker pull`` to every Determined agent machine in the cluster. Users can
configure different container images for NVIDIA GPU tasks using the ``cuda`` key (``gpu`` prior to
Determined 0.17.6), CPU tasks using ``cpu`` key, and ROCm (AMD GPU) tasks using the ``rocm`` key.
Default values:

-  ``determinedai/pytorch-ngc-dev:f20b027`` for NVIDIA GPUs and for CPUs.
-  ``determinedai/environments:rocm-5.0-pytorch-1.10-tf-2.7-rocm-0.26.4`` for ROCm.

For TensorFlow users, we provide an image that must be referenced in the experiment configuration:

-  ``determinedai/tensorflow-ngc-dev:f20b027`` for NVIDIA GPUs and for CPUs.

``environment_variables``
=========================

A list of environment variables that will be set in every task container. Each element of the list
should be a string of the form ``NAME=VALUE``. See :ref:`environment-variables` for more details.
Environment variables specified in the experiment configuration will override default values
specified here. You can customize environment variables for CUDA (NVIDIA GPU), CPU, and ROCm (AMD
GPU) tasks differently by specifying a dict with ``cuda`` (``gpu`` prior to Determined 0.17.6),
``cpu``, and ``rocm`` keys.

``startup_hook``
================

An optional inline script that will be executed as part of task set up. This is defined under
`task_container_defaults` at master or resource pool level. This script will be executed using
`/bin/bash`.

``log_policies``
================

A list of log policies that take effect when a trial reports a log that matches a pattern. For
details, visit :ref:`log_policies <config-log-policies>`.

``force_pull_image``
====================

Defines the default policy for forcibly pulling images from the Docker registry and bypassing the
Docker cache. If a pull policy is specified in the :ref:`experiment configuration
<exp-environment-image>` this default value is overridden. Please note that as of November 1st,
2020, unauthenticated users will be `capped at 100 pulls from Docker Hub per 6 hours
<https://www.docker.com/blog/scaling-docker-to-serve-millions-more-developers-network-egress/>`__.
Defaults to ``false``.

``registry_auth``
=================

Defines the default `Docker registry credentials
<https://docs.docker.com/engine/api/v1.30/#tag/System/operation/SystemAuth>`__ to use when pulling a
custom base Docker image, if needed. If credentials are specified in the :ref:`experiment config
<exp-environment-image>` this default value is overridden. Credentials are specified as the
following nested fields:

-  ``username`` (required)
-  ``password`` (required)
-  ``serveraddress`` (required)
-  ``email`` (optional)

``add_capabilities``
====================

The default list of Linux capabilities to grant to task containers. Ignored by resource managers of
type ``kubernetes``. See :ref:`environment.add_capabilities <exp-environment-add-capabilities>` for
more details.

``drop_capabilities``
=====================

Just like ``add_capabilities`` but for dropping capabilities.

``devices``
===========

The default list of devices to pass to the Docker daemon. Ignored by resource managers of type
``kubernetes``. See :ref:`resources.devices <exp-resources-devices>` for more details.

``bind_mounts``
===============

The default bind mounts to pass to the Docker container. Ignored by resource managers of type
``kubernetes``. See :ref:`bind_mounts <exp-bind-mounts>` for more details.

``kubernetes``
==============

``max_slots_per_pod`` See :ref:`resource_manager.max_slots
<master-config-reference-max-slots-per-pod>` for more details.

``slurm``
=========

Additional Slurm options when launching trials with ``sbatch``. See :ref:`environment.slurm
<slurm-config>` for more details.

``pbs``
=======

Additional PBS options when launching trials with ``qsub``. See :ref:`environment.pbs <pbs-config>`
for more details.

**********
 ``root``
**********

Specifies the root directory of the state files. Defaults to ``/usr/share/determined/master``.

***********
 ``cache``
***********

Configuration for file cache.

``cache_dir``
=============

Specifies the root directory for file cache. Defaults to ``/var/cache/determined``. Note that the
master would break on startup if it does not have access to create this default directory.

******************
 ``launch_error``
******************

Optional. Specifies whether to refuse an experiment or task if the slots requested exceeds the
cluster capacity. This option has no effect for Kubernetes or Slurm clusters. If ``false``, only a
warning is returned. The default value is ``true``.

******************
 ``cluster_name``
******************

Optional. Specify a human-readable name for this cluster.

**********************
 ``ui_customization``
**********************

Optional. Applies only to the Determined Enterprise Edition. This section houses options to
customize the UI.

``logo_path``
=============

Specifies the path to a user-provided logo to be shown in the UI. Ensure the path is accessible and
reachable by the master service. The logo file should be a valid image format, with SVG recommended.

*************************
 ``tensorboard_timeout``
*************************

Specifies the duration in seconds before idle TensorBoard instances are automatically terminated. A
TensorBoard instance is considered to be idle if it does not receive any HTTP traffic. The default
timeout is ``300`` (5 minutes).

.. _master-config-notebook-timeout:

**********************
 ``notebook_timeout``
**********************

Specifies the duration in seconds before idle notebook instances are automatically terminated. A
notebook instance is considered to be idle if it is not receiving any HTTP traffic and it is not
otherwise active (as defined by the ``notebook_idle_type`` option in the :ref:`task configuration
<command-notebook-configuration>`). Defaults to ``null``, i.e. disabled.

.. _master-config-resource-manager:

**********************
 ``resource_manager``
**********************

The resource manager used to acquire resources. Defaults to ``agent``.

For Kubernetes installations, if you define additional resource managers, the resource manager
specified under the primary resource_manager key here is considered the default.

``name``
========

Optional. Specifies the resource manager's name. Defaults to ``default`` if not specified. For
Kubernetes installations with additional resource managers, ensure unique names for all resource
managers in the cluster.

``metadata``
============

Optional. Stores additional information about the resource manager in a yaml map, such as the zone,
region, or location.

For example:

.. code:: yaml

   metadata:
      region: us-west1
      zone: us-west1-a

``type: agent``
===============

The agent resource manager includes static and dynamic agents.

``scheduler``
-------------

Specifies how Determined schedules tasks to agents on resource pools. If a resource pool is
specified with an individual scheduler configuration, that will override the default scheduling
behavior specified here. For more on scheduling behavior in Determined, see :ref:`scheduling`.

``type``
^^^^^^^^

   The scheduling policy to use when allocating resources between different tasks (experiments,
   notebooks, etc.). Defaults to ``priority``.

   -  ``fair_share``: (deprecated) Tasks receive a proportional amount of the available resources
      depending on the resource they require and their weight.

   -  ``priority``: Tasks are scheduled based on their priority, which can range from the values 1
      to 99 inclusive. Lower priority numbers indicate higher-priority tasks. A lower-priority task
      will never be scheduled while a higher-priority task is pending. Zero-slot tasks (e.g.,
      CPU-only notebooks, TensorBoards) are prioritized separately from tasks requiring slots (e.g.,
      experiments running on GPUs). Task priority can be assigned using the ``resources.priority``
      field. If a task does not specify a priority it is assigned the ``default_priority``.

      -  ``preemption``: Specifies whether lower-priority tasks should be preempted to schedule
         higher priority tasks. Tasks are preempted in order of lowest priority first.
      -  ``default_priority``: The priority that is assigned to tasks that do not specify a
         priority. Can be configured to 1 to 99 inclusively. Defaults to ``42``.

``fitting_policy``
^^^^^^^^^^^^^^^^^^

   The scheduling policy to use when assigning tasks to agents in the cluster. Defaults to ``best``.

   -  ``best``: The best-fit policy ensures that tasks will be preferentially "packed" together on
      the smallest number of agents.
   -  ``worst``: The worst-fit policy ensures that tasks will be placed on under-utilized agents.

.. _allow-uneven-slots:

``allow_heterogeneous_fits``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

   Fit distributed jobs onto agents of different sizes. When enabled, we still prefer to fit jobs on
   same sized nodes but will fallback to allow heterogeneous fits. Sizes should be powers of two for
   the fitting algorithm to work.

``default_aux_resource_pool``
-----------------------------

The default resource pool to use for tasks that do not need dedicated compute resources, auxiliary,
or systems tasks. Defaults to ``default`` if no resource pool is specified.

``default_compute_resource_pool``
---------------------------------

The default resource pool to use for tasks that require compute resources, e.g. GPUs or dedicated
CPUs. Defaults to ``default`` if no resource pool is specified.

``require_authentication``
--------------------------

Whether to require that agent connections be verified using mutual TLS.

``client_ca``
-------------

Certificate authority file to use for verifying agent certificates.

``type: kubernetes``
====================

The ``kubernetes`` resource manager launches tasks on a Kubernetes cluster. The Determined master
must be running within the Kubernetes cluster. When using the ``kubernetes`` resource manager, we
recommend deploying Determined using the :ref:`Determined Helm Chart <install-on-kubernetes>`. When
installed via Helm, the configuration settings below will be set automatically. For more information
on using Determined with Kubernetes, see the :ref:`documentation <determined-on-kubernetes>`.

``namespace``
-------------

The namespace where Determined will deploy Pods and ConfigMaps.

.. _master-config-reference-max-slots-per-pod:

``max_slots_per_pod``
---------------------

Each multi-slot (distributed training) task will be scheduled as a set of ``slots_per_task /
max_slots_per_pod`` separate pods, with each pod assigned up to ``max_slots_per_pod`` slots.
Distributed tasks with sizes that are not divisible by ``max_slots_per_pod`` are never scheduled. If
you have a cluster of different size nodes, set ``max_slots_per_pod`` to the greatest common divisor
of all the sizes. For example, if you have some nodes with 4 GPUs and other nodes with 8 GPUs, set
``maxSlotsPerPod`` to ``4`` so that all distributed experiments will launch with 4 GPUs per pod
(with two pods on 8-GPU nodes).

This field can also be set in ``task_container_defaults.kubernetes.max_slots_per_pod`` to allow per
resource pool ``max_slots_per_pod``.

``slot_type``
-------------

Resource type used for compute tasks. Defaults to ``cuda``.

``slot_type: cuda``
^^^^^^^^^^^^^^^^^^^

   One NVIDIA GPU will be requested per compute slot. Prior to Determined 0.17.6, this option was
   called ``gpu``.

``slot_type: cpu``
^^^^^^^^^^^^^^^^^^

   CPU resources will be requested for each compute slot. ``slot_resource_requests.cpu`` option is
   required to specify the specific amount of the resources.

``slot_resource_requests``
--------------------------

Supports customizing the resource requests made when scheduling Kubernetes pods.

``cpu``
^^^^^^^

   The number of Kubernetes CPUs to request per compute slot.

``master_service_name``
-----------------------

The service account Determined uses to interact with the Kubernetes API.

.. _cluster-configuration-slurm:

``type: slurm`` or ``pbs``
==========================


master/internal/config/config.go
package config

import (
	"crypto/rand"
	"crypto/tls"
	"encoding/hex"
	"encoding/json"
	"encoding/pem"
	"fmt"
	"net/url"
	"path/filepath"
	"sync"

	"github.com/jinzhu/copier"
	"github.com/pkg/errors"
	log "github.com/sirupsen/logrus"

	"github.com/determined-ai/determined/master/pkg/config"
	pkgconfig "github.com/determined-ai/determined/master/pkg/config"
	"github.com/determined-ai/determined/master/pkg/model"
	"github.com/determined-ai/determined/master/pkg/schemas/expconf"
)

// These are package-level variables so that they can be set at link time.
// WARN: if you move them to a different package, you need to change the linked
// path in the make file and CI.
var (
	DefaultSegmentMasterKey = ""
	DefaultSegmentWebUIKey  = ""
)

var (
	once         sync.Once
	masterConfig *Config
)

// KubernetesDefaultPriority is the default K8 resource manager priority.
const (
	KubernetesDefaultPriority = 50
	sslModeDisable            = "disable"
)

type (
	// ExperimentConfigPatch is the updatedble fields for patching an experiment.
	ExperimentConfigPatch struct {
		Name *string `json:"name,omitempty"`
	}
)

// DefaultDBConfig returns the default configuration of the database.
func DefaultDBConfig() *DBConfig {
	return &DBConfig{
		Migrations: "file://static/migrations",
		SSLMode:    sslModeDisable,
	}
}

// CacheConfig is the configuration for file cache.
type CacheConfig struct {
	CacheDir string `json:"cache_dir"`
}

// DBConfig hosts configuration fields of the database.
type DBConfig struct {
	User             string `json:"user"`
	Password         string `json:"password"`
	Migrations       string `json:"migrations"`
	ViewsAndTriggers string `json:"views_and_triggers"`
	Host             string `json:"host"`
	Port             string `json:"port"`
	Name             string `json:"name"`
	SSLMode          string `json:"ssl_mode"`
	SSLRootCert      string `json:"ssl_root_cert"`
}

// WebhooksConfig hosts configuration fields for webhook functionality.
type WebhooksConfig struct {
	BaseURL    string `json:"base_url"`
	SigningKey string `json:"signing_key"`
}

// IntegrationsConfig stores configs related to integrations like pachyderm.
type IntegrationsConfig struct {
	Pachyderm PachydermConfig `json:"pachyderm"`
}

// PachydermConfig stores fields needed to integrate Pachyderm with determined.
type PachydermConfig struct {
	// pachyderm daemon address.
	Address string `json:"address"`
}

// DefaultConfig returns the default configuration of the master.
func DefaultConfig() *Config {
	return &Config{
		ConfigFile:            "",
		Log:                   *pkgconfig.DefaultLoggerConfig(),
		DB:                    *DefaultDBConfig(),
		TaskContainerDefaults: *model.DefaultTaskContainerDefaults(),
		TensorBoardTimeout:    5 * 60,
		Security: SecurityConfig{
			DefaultTask: model.AgentUserGroup{
				UID:   0,
				GID:   0,
				User:  "root",
				Group: "root",
			},
			SSH: SSHConfig{
				RsaKeySize: 1024,
			},
			AuthZ: *DefaultAuthZConfig(),
		},
		// If left unspecified, the port is later filled in with 8080 (no TLS) or 8443 (TLS).
		Port: 0,
		Root: "/usr/share/determined/master",
		Telemetry: config.TelemetryConfig{
			Enabled:                  true,
			OtelEnabled:              false,
			OtelExportedOtlpEndpoint: "localhost:4317",
			SegmentMasterKey:         DefaultSegmentMasterKey,
			SegmentWebUIKey:          DefaultSegmentWebUIKey,
		},
		EnableCors:  false,
		LaunchError: true,
		ClusterName: "",
		Logging:     model.LoggingConfig{DefaultLoggingConfig: &model.DefaultLoggingConfig{}},
		// For developers this should be a writable directory for caching files.
		Cache: CacheConfig{
			CacheDir: "/var/cache/determined",
		},
		FeatureSwitches: []string{},
		ResourceConfig:  *DefaultResourceConfig(),
		Observability: ObservabilityConfig{
			EnablePrometheus: true,
		},
		OIDC: OIDCConfig{
			AuthenticationClaim:         "email",
			SCIMAuthenticationAttribute: "userName",
			AutoProvisionUsers:          false,
		},
	}
}

// Config is the configuration of the master.
//
// It is populated, in the following order, by the master configuration file,
// environment variables and command line arguments.
type Config struct {
	ConfigFile            string                            `json:"config_file"`
	Log                   pkgconfig.LoggerConfig            `json:"log"`
	DB                    DBConfig                          `json:"db"`
	TensorBoardTimeout    int                               `json:"tensorboard_timeout"`
	NotebookTimeout       *int                              `json:"notebook_timeout"`
	Security              SecurityConfig                    `json:"security"`
	CheckpointStorage     expconf.CheckpointStorageConfig   `json:"checkpoint_storage"`
	TaskContainerDefaults model.TaskContainerDefaultsConfig `json:"task_container_defaults"`
	Port                  int                               `json:"port"`
	Root                  string                            `json:"root"`
	Telemetry             config.TelemetryConfig            `json:"telemetry"`
	EnableCors            bool                              `json:"enable_cors"`
	LaunchError           bool                              `json:"launch_error"`
	ClusterName           string                            `json:"cluster_name"`
	Logging               model.LoggingConfig               `json:"logging"`
	RetentionPolicy       model.LogRetentionPolicy          `json:"retention_policy"`
	Observability         ObservabilityConfig               `json:"observability"`
	Cache                 CacheConfig                       `json:"cache"`
	Webhooks              WebhooksConfig                    `json:"webhooks"`
	FeatureSwitches       []string                          `json:"feature_switches"`
	ReservedPorts         []int                             `json:"reserved_ports"`
	ResourceConfig

	// Internal contains "hidden" useful debugging configurations.
	InternalConfig InternalConfig `json:"__internal"`

	Scim         ScimConfig         `json:"scim"`
	SAML         SAMLConfig         `json:"saml"`
	OIDC         OIDCConfig         `json:"oidc"`
	DetCloud     DetCloudConfig     `json:"det_cloud"`
	Integrations IntegrationsConfig `json:"integrations"`
}

// GetMasterConfig returns reference to the master config singleton.
func GetMasterConfig() *Config {
	once.Do(func() {
		masterConfig = DefaultConfig()
	})
	return masterConfig
}

// SetMasterConfig sets the master config singleton.
func SetMasterConfig(aConfig *Config) {
	if masterConfig != nil {
		panic("master config is already set")
	}
	if aConfig == nil {
		panic("passed in config is nil")
	}
	config := GetMasterConfig()
	*config = *aConfig
}

// Printable returns a printable string.
func (c Config) Printable() ([]byte, error) {
	var configCopy Config
	if err := copier.CopyWithOption(
		&configCopy, &c, copier.Option{DeepCopy: true, IgnoreEmpty: true},
	); err != nil {
		return nil, fmt.Errorf("copying config: %w", err)
	}

	const hiddenValue = "********"
	if configCopy.Security.InitialUserPassword != "" {
		configCopy.Security.InitialUserPassword = hiddenValue
	}
	if configCopy.DB.Password != "" {
		configCopy.DB.Password = hiddenValue
	}
	if configCopy.Telemetry.SegmentMasterKey != "" {
		configCopy.Telemetry.SegmentMasterKey = hiddenValue
	}
	if configCopy.Telemetry.SegmentWebUIKey != "" {
		configCopy.Telemetry.SegmentWebUIKey = hiddenValue
	}
	if configCopy.TaskContainerDefaults.RegistryAuth != nil {
		if configCopy.TaskContainerDefaults.RegistryAuth.Password != "" {
			// RegistryAuth is a pointer, so if we need to hide the password we need to be very
			// careful to replace the pointer, not the contents behind the pointer.
			printable := *configCopy.TaskContainerDefaults.RegistryAuth
			printable.Password = hiddenValue
			configCopy.TaskContainerDefaults.RegistryAuth = &printable
		}
	}

	// When there are pointers inside the type, we need to copy things to avoid modifying the original
	// object.
	if origAuth := c.Scim.Auth.BasicAuthConfig; origAuth != nil {
		auth := *origAuth
		auth.Username = hiddenValue
		auth.Password = hiddenValue
		c.Scim.Auth.BasicAuthConfig = &auth
	}

	configCopy.CheckpointStorage = configCopy.CheckpointStorage.Printable()

	maskPools := func(pools []ResourcePoolConfig) []ResourcePoolConfig {
		for i, p := range pools {
			pools[i] = p.Printable()
		}
		return pools
	}
	configCopy.RootPoolsInternal = maskPools(configCopy.RootPoolsInternal)
	for _, r := range configCopy.AdditionalResourceManagersInternal {
		r.ResourcePools = maskPools(r.ResourcePools)
	}

	optJSON, err := json.Marshal(configCopy)
	if err != nil {
		return nil, errors.Wrap(err, "unable to convert config to JSON")
	}
	return optJSON, nil
}

func k8sValidateMaxSlots(r *ResourceManagerWithPoolsConfig,
	taskContainerDefaults model.TaskContainerDefaultsConfig, totalRMs int,
) (model.TaskContainerDefaultsConfig, error) {
	if taskContainerDefaults.Kubernetes == nil {
		taskContainerDefaults.Kubernetes = &model.KubernetesTaskContainerDefaults{}
	}

	rmMaxSlots := r.ResourceManager.KubernetesRM.MaxSlotsPerPod
	taskMaxSlots := taskContainerDefaults.Kubernetes.MaxSlotsPerPod

	// if exactly one resource manager, allow global task default to be used
	if totalRMs == 1 {
		if (rmMaxSlots != nil) == (taskMaxSlots != nil) {
			return taskContainerDefaults, fmt.Errorf("must provide exactly one of " +
				"resource_manager.max_slots_per_pod and " +
				"task_container_defaults.kubernetes.max_slots_per_pod")
		}

		if rmMaxSlots != nil {
			taskContainerDefaults.Kubernetes.MaxSlotsPerPod = rmMaxSlots
		}
		if taskMaxSlots != nil {
			r.ResourceManager.KubernetesRM.MaxSlotsPerPod = taskMaxSlots
		}
	} else {
		// otherwise, must use max slots defined in resource manager config
		if rmMaxSlots == nil {
			return taskContainerDefaults, fmt.Errorf("must provide resource_manager.max_slots_per_pod")
		}
		if taskMaxSlots != nil {
			log.Warn("ignoring task_container_defaults.kubernetes.max_slots_per_pod - " +
				"must provide resource_manager.max_slots_per_pod " +
				"if multiple resource managers are defined")
		}
	}

	if maxSlotsPerPod := *r.ResourceManager.KubernetesRM.MaxSlotsPerPod; maxSlotsPerPod < 0 {
		return taskContainerDefaults, fmt.Errorf("max_slots_per_pod must be >= 0 got %d", maxSlotsPerPod)
	}
	return taskContainerDefaults, nil
}

// Resolve resolves the values in the configuration.
func (c *Config) Resolve() error {
	if c.Port == 0 {
		if c.Security.TLS.Enabled() {
			c.Port = 8443
		} else {
			c.Port = 8080
		}
	}

	root, err := filepath.Abs(c.Root)
	if err != nil {
		return err
	}
	c.Root = root

	c.DB.Migrations = fmt.Sprintf("file://%s", filepath.Join(c.Root, "static/migrations"))
	c.DB.ViewsAndTriggers = filepath.Join(c.Root, "static/views_and_triggers")

	// We must resolve resources before we apply pool defaults.
	if err := c.ResolveResource(); err != nil {
		return err
	}

	for _, r := range c.ResourceManagers() {
		if r.ResourceManager.AgentRM != nil {
			if r.ResourceManager.AgentRM.Scheduler == nil {
				r.ResourceManager.AgentRM.Scheduler = DefaultSchedulerConfig()
			}
			if r.ResourceManager.AgentRM.Scheduler.GetType() == FairShareScheduling {
				log.Warn("Fair-Share Scheduler has been deprecated, please update master config to use Priority Scheduler.")
			}
			if r.ResourceManager.AgentRM.Scheduler.GetType() == RoundRobinScheduling {
				log.Error("Round Robin Scheduler has been removed, please update master config to use Priority Scheduler.")
				log.Info("Priority Scheduler with all priorities equal will have the same behavior as a Round Robin Scheduler.")
				return fmt.Errorf("scheduler not available")
			}
		}

		if r.ResourceManager.KubernetesRM != nil {
			c.TaskContainerDefaults, err = k8sValidateMaxSlots(r, c.TaskContainerDefaults, len(c.ResourceManagers()))
			if err != nil {
				return err
			}
		}
	}

	if c.Webhooks.SigningKey == "" {
		b := make([]byte, 6)
		if _, err := rand.Read(b); err != nil {
			return err
		}
		c.Webhooks.SigningKey = hex.EncodeToString(b)
	}

	if err := c.Logging.Resolve(); err != nil {
		return err
	}

	if c.Security.AuthZ.StrictNTSCEnabled {
		log.Warn("_strict_ntsc_enabled option is removed and will not have any effect.")
	}

	if c.OIDC.AutoProvisionUsers && c.Scim.Enabled {
		log.Warn("scim enabled; overriding OIDC user & group provisions")
		c.OIDC.AutoProvisionUsers = false
		c.OIDC.GroupsAttributeName = ""
	}

	if c.OIDC.GroupsAttributeName != "" && !c.Security.AuthZ.IsRBACUIEnabled() {
		log.Warn("groups_attribute_name requires rbac to be enabled")
		c.OIDC.GroupsAttributeName = ""
	}

	if c.SAML.AutoProvisionUsers && c.Scim.Enabled {
		log.Warn("scim enabled; overriding SAML user & group provisions")
		c.SAML.AutoProvisionUsers = false
		c.SAML.GroupsAttributeName = ""
	}

	if c.SAML.GroupsAttributeName != "" && !c.Security.AuthZ.IsRBACUIEnabled() {
		log.Warn("groups_attribute_name requires rbac to be enabled")
		c.SAML.GroupsAttributeName = ""
	}

	return nil
}

// Deprecations describe fields which were recently or will soon be removed.
func (c *Config) Deprecations() (errs []error) {
	for _, r := range c.ResourceManagers() {
		for _, rp := range r.ResourcePools {
			if rp.AgentReattachEnabled {
				errs = append(errs, fmt.Errorf(
					"agent_reattach_enabled is set for resource pool %s but will be ignored; "+
						"as of 0.21.0 this feature is always on", rp.PoolName,
				))
			}
		}
	}
	return errs
}

// SecurityConfig is the security configuration for the master.
type SecurityConfig struct {
	DefaultTask model.AgentUserGroup `json:"default_task"`
	TLS         TLSConfig            `json:"tls"`
	SSH         SSHConfig            `json:"ssh"`
	AuthZ       AuthZConfig          `json:"authz"`

	InitialUserPassword string `json:"initial_user_password"`
}

// SSHConfig is the configuration setting for SSH.
type SSHConfig struct {
	RsaKeySize int `json:"rsa_key_size"`
}

// TLSConfig is the configuration for setting up serving over TLS.
type TLSConfig struct {
	Cert string `json:"cert"`
	Key  string `json:"key"`
}

// Validate implements the check.Validatable interface.
func (t *TLSConfig) Validate() []error {
	var errs []error
	if t.Cert == "" && t.Key != "" {
		errs = append(errs, errors.New("TLS key file provided without a cert file"))
	} else if t.Key == "" && t.Cert != "" {
		errs = append(errs, errors.New("TLS cert file provided without a key file"))
	}
	return errs
}

// Validate implements the check.Validatable interface.
func (t *SSHConfig) Validate() []error {
	var errs []error
	if t.RsaKeySize < 1 {
		errs = append(errs, errors.New("RSA Key size must be greater than 0"))
	} else if t.RsaKeySize > 16384 {
		errs = append(errs, errors.New("RSA Key size must be less than 16,384"))
	}
	return errs
}

// Enabled returns whether this configuration makes it possible to enable TLS.
func (t *TLSConfig) Enabled() bool {
	return t.Cert != "" && t.Key != ""
}

// ReadCertificate returns the certificate described by this configuration (nil if it does not allow
// TLS to be enabled).
func (t *TLSConfig) ReadCertificate() (*tls.Certificate, error) {
	if !t.Enabled() {
		return nil, nil
	}
	cert, err := tls.LoadX509KeyPair(t.Cert, t.Key)
	return &cert, err
}

// ProxiedServerConfig is the configuration for a internal proxied server.
type ProxiedServerConfig struct {
	// Prefix is the path prefix to match for this proxy.
	PathPrefix string `json:"path_prefix"`
	// Destination is the URL to proxy to.
	Destination string `json:"destination"`
}

// InternalConfig is the configuration for internal knobs.
type InternalConfig struct {
	AuditLoggingEnabled bool                   `json:"audit_logging_enabled"`
	ExternalSessions    model.ExternalSessions `json:"external_sessions"`
	ProxiedServers      []ProxiedServerConfig  `json:"proxied_servers"`
	PreemptionTimeout   *model.Duration        `json:"preemption_timeout"`
}

// Validate implements the check.Validatable interface.
func (p *ProxiedServerConfig) Validate() []error {
	var errs []error
	if p.PathPrefix == "" {
		errs = append(errs, errors.New("path_prefix must be set"))
	}
	if p.Destination == "" {
		errs = append(errs, errors.New("destination must be set"))
		return errs
	}
	target, err := url.Parse(p.Destination)
	if err != nil {
		errs = append(errs, errors.Wrap(err, "failed to parse proxied destination"))
		return errs
	}
	// ensure scheme and port is set
	if target.Scheme == "" {
		target.Scheme = "http"
	}
	if target.Port() == "" {
		errs = append(errs, errors.New("proxy path must include a port"))
	}
	return errs
}

// Validate implements the check.Validatable interface.
func (i *InternalConfig) Validate() []error {
	var errs []error
	// We allow setting multiple proxied servers but leave it up to the developer
	// to ensure that they don't conflict with eachother or other det routes.
	for _, p := range i.ProxiedServers {
		errs = append(errs, p.Validate()...)
	}
	return errs
}

// ObservabilityConfig is the configuration for observability metrics.
// Defaulted to true.
type ObservabilityConfig struct {
	EnablePrometheus bool `json:"enable_prometheus"`
}

func readPriorityFromScheduler(conf *SchedulerConfig) *int {
	if conf == nil || conf.Priority == nil {
		return nil
	}
	return conf.Priority.DefaultPriority
}

// ReadRMPreemptionStatus resolves the preemption status for a resource manager.
// TODO(Brad): Move these to a resource pool level API.
func ReadRMPreemptionStatus(rpName string) bool {
	config := GetMasterConfig()

	for _, r := range config.ResourceManagers() {
		for _, rpConfig := range r.ResourcePools {
			if rpConfig.PoolName == rpName {
				if rpConfig.Scheduler != nil {
					return rpConfig.Scheduler.GetPreemption()
				}
				// if not found, fall back to resource manager config
				return readRMPreemptionStatus(r, rpName)
			}
		}
	}

	// if not found in any RMs, return default RM config
	if len(config.ResourceManagers()) != 0 {
		return readRMPreemptionStatus(config.ResourceManagers()[0], rpName)
	}
	panic("unexpected resource configuration")
}

func readRMPreemptionStatus(config *ResourceManagerWithPoolsConfig, rpName string) bool {
	for _, rpConfig := range config.ResourcePools {
		if rpConfig.PoolName != rpName {
			continue
		}
		if rpConfig.Scheduler != nil {
			return rpConfig.Scheduler.GetPreemption()
		}
		break
	}

	// if not found, fall back to resource manager config
	switch {
	case config.ResourceManager.AgentRM != nil:
		if config.ResourceManager.AgentRM.Scheduler == nil {
			panic("scheduler not configured")
		}
		return config.ResourceManager.AgentRM.Scheduler.GetPreemption()
	case config.ResourceManager.KubernetesRM != nil:
		return config.ResourceManager.KubernetesRM.GetPreemption()
	case config.ResourceManager.DispatcherRM != nil,
		config.ResourceManager.PbsRM != nil:
		return false
	default:
		panic("unexpected resource configuration")
	}
}

// ReadPriority resolves the priority value for a job.
func ReadPriority(rpName string, jobConf interface{}) int {
	config := GetMasterConfig()
	var prio *int
	// look at the individual job config
	switch conf := jobConf.(type) {
	case *expconf.ExperimentConfig:
		prio = conf.Resources().Priority()
	case *model.CommandConfig:
		prio = conf.Resources.Priority
	}
	if prio != nil {
		return *prio
	}

	// if not found, fall back to the resource pools config
	for _, rm := range config.ResourceManagers() {
		for _, rpConfig := range rm.ResourcePools {
			if rpConfig.PoolName == rpName {
				schedulerConf := rpConfig.Scheduler
				prio = readPriorityFromScheduler(schedulerConf)
				if prio != nil {
					return *prio
				}

				// if not found, fall back to resource manager config
				if rm.ResourceManager.AgentRM != nil {
					schedulerConf = rm.ResourceManager.AgentRM.Scheduler
					prio = readPriorityFromScheduler(schedulerConf)
					if prio != nil {
						return *prio
					}
				}

				if rm.ResourceManager.KubernetesRM != nil {
					return KubernetesDefaultPriority
				}

				break
			}
		}
	}

	return DefaultSchedulingPriority
}

// ReadWeight resolves the weight value for a job.
func ReadWeight(rpName string, jobConf interface{}) float64 {
	var weight float64
	switch conf := jobConf.(type) {
	case *expconf.ExperimentConfig:
		weight = conf.Resources().Weight()
	case *model.CommandConfig:
		weight = conf.Resources.Weight
	}
	return weight
}

// GetCertPEM returns the PEM-encoded certificate.
func GetCertPEM(cert *tls.Certificate) []byte {
	var certBytes []byte
	if cert != nil {
		for _, c := range cert.Certificate {
			b := pem.EncodeToMemory(&pem.Block{
				Type:  "CERTIFICATE",
				Bytes: c,
			})
			certBytes = append(certBytes, b...)
		}
	}
	return certBytes
}
